_studio/mfx_lib/encode_hw/h264/src/mfx_h264_encode_hw.cpp

```diff
@@ -2067,29 +2082,38 @@ void ImplementationAvc::BrcPreEnc(
 
     m_brc.PreEnc(task.m_brcFrameParams, m_tmpVmeData);
 }
-#if defined(MXF_ENABLE_MCTF_IN_AVC)
-mfxStatus ImplementationAvc::SubmitToMctf(DdiTask * pTask, bool isSceneChange, bool doIntraFiltering)
+#if defined(MFX_ENABLE_MCTF_IN_AVC)
+mfxStatus ImplementationAvc::SubmitToMctf(DdiTask * pTask)
 {
     MFX_AUTO_LTRACE(MFX_TRACE_LEVEL_API, "VideoVPPHW::SubmitToMctf");
-    //mfxStatus sts = MFX_ERR_NONE;
-
-    pTask->m_bFrameReady   = false;
-    mfxFrameSurface1
-        * pSurfI         = nullptr;
-    pSurfI               = m_core->GetNativeSurface(pTask->m_yuv);
-    pSurfI               = pSurfI ? pSurfI : pTask->m_yuv;
 
-    amtMctf->IntBufferUpdate(isSceneChange);
+    pTask->m_bFrameReady = false;
+    bool isIntraFrame    = (pTask->GetFrameType() & MFX_FRAMETYPE_I) || (pTask->GetFrameType() & MFX_FRAMETYPE_IDR);
+    bool isPFrame        = (pTask->GetFrameType() & MFX_FRAMETYPE_P);
+    bool isAnchorFrame   = isIntraFrame || isPFrame;
 
+    amtMctf->IntBufferUpdate(pTask->m_SceneChange, isIntraFrame, pTask->m_doMCTFIntraFiltering);
+    if (isAnchorFrame || pTask->m_SceneChange)
+    {
+        pTask->m_idxMCTF = FindFreeResourceIndex(m_mctf);
+        pTask->m_midMCTF = AcquireResource(m_mctf, pTask->m_idxMCTF);
+        if (pTask->m_midMCTF)
+        {
+            MFX_SAFE_CALL(m_core->GetFrameHDL(pTask->m_midMCTF, &pTask->m_handleMCTF.first));
+        }
+        else
+            isAnchorFrame = false; //No resource available to generate filtered output, let it pass.
+    }
     if (IsCmNeededForSCD(m_video))
     {
-        mfxStatus sts = GetNativeHandleToRawSurface(*m_core, m_video, *pTask, pTask->m_handleRaw);
-        if (sts != MFX_ERR_NONE)
-            return Error(sts);
-
-        if(amtMctf->MCTF_Check_Use())
-            pTask->m_cmRawForMCTF = CreateSurface(m_cmDevice, pTask->m_handleRaw, m_currentVaType);
-        MFX_SAFE_CALL(amtMctf->MCTF_PUT_FRAME(pTask->m_cmRawForMCTF, isSceneChange, true, nullptr, doIntraFiltering));
+        MFX_SAFE_CALL(amtMctf->MCTF_PUT_FRAME(
+            pTask->m_yuv,
+            pTask->m_handleMCTF,
```

Comment:
what if block at 2102 was not executed? Is additional check for 0 needed for pTask->m_handleMCTF?

Reply:
In my opinion there is no need for that, all other functions are taking care of it.